src/cpp/fastdds/builtin/type_lookup_service/TypeLookupReplyListener.hpp

```diff
@@ -0,0 +1,183 @@
+
+    /**
+     * @brief Constructor
+     * @param manager Pointer to the TypeLookupManager.
+     */
+    TypeLookupReplyListener(
+            TypeLookupManager* manager);
+
+    /**
+     * @brief Destructor
+     */
+    virtual ~TypeLookupReplyListener() override;
+
+protected:
+
+    /**
+     * @brief Starts the thread that process the received replies.
+     */
+    void start_reply_processor_thread();
+
+    /**
+     * @brief Stops the thread that process the received replies.
+     */
+    void stop_reply_processor_thread();
+
+    /**
+     * @brief Process the replies in the queue.
+     */
+    void process_reply();
+
+    /**
+     * @brief Registers TypeIdentifier and TypeObject in TypeObjectRegistry.
+     * This method also notifies all type related callbacks and removes the current SampleIdentity from the pending request list.
+     * @param request_id[in] The SampleIdentity of the request.
+     * @param reply[in] The reply data.
+     * @param related_request[in] The request that this reply answers.
+     */
+    void check_get_types_reply(
+            const SampleIdentity& request_id,
+            const TypeLookup_getTypes_Out& reply,
+            SampleIdentity related_request);
+
+    /**
+     * @brief Checks if all dependencies are solved.
+     * If they are not, sends next request and adds it to the list.
+     * If they are, sends get_types request and adds it to the list.
+     * Also removes the current SampleIdentity from the list.
+     * @param request_id[in] The SampleIdentity of the request.
+     * @param type_server[in] GUID corresponding to the remote participant which TypeInformation is being solved.
+     * @param reply[in] The reply data.
+     */
+    void check_get_type_dependencies_reply(
+            const SampleIdentity& request_id,
+            const fastrtps::rtps::GUID_t type_server,
+            const TypeLookup_getTypeDependencies_Out& reply);
+
+    /**
+     * @brief Method called when this class is notified of a new cache change.
+     * @param reader The reader receiving the cache change.
+     * @param change The cache change.
+     */
+    void onNewCacheChangeAdded(
+            fastrtps::rtps::RTPSReader* reader,
+            const fastrtps::rtps::CacheChange_t* const change) override;
+
+
+    //! A pointer to the typelookup manager.
+    TypeLookupManager* typelookup_manager_;
+
+    //! Mutex to protect access to replies_with_continuation_.
+    std::mutex replies_with_continuation_mutex_;
+
+    //! Collection of the replies that needed continuation points.
+    std::vector<SampleIdentity> replies_with_continuation_;
+
+    eprosima::thread replies_processor_thread;
+    std::queue<ReplyWithServerGUID> replies_queue_;
+    std::mutex replies_processor_cv_mutex_;
+    std::condition_variable replies_processor_cv_;
+    bool processing_;
+    rtps::ThreadSettings replies_processor_thread_settings_;
+};
+
+class TypeLookupReplyWListener : public fastrtps::rtps::WriterListener
```

Comment:
Instead of having a second class, I suggest having only one listener class inheriting from both Writer and Reader Listener. Same for the Request listener

Reply:
removed class in [8cefffb](https://github.com/eProsima/Fast-DDS/pull/4240/commits/8cefffbf00f1b497a9bd31bcdb75802ef4bdbc3a)